amworkflow/gcode/gcode.py

```diff
@@ -35,7 +39,263 @@ def create(self, in_file: Path, out_gcode: Path) -> None:
+                hex_values = [hex(int(''.join(map(str, chunk.astype(int))), 2))[2:].upper() for chunk in chunks]
+
+                # Combine hex numbers
+                printer_hex.append(hex_values)
+
+            with open(out_dsmn, "a") as file:
+                for hex_values in printer_hex:
+                    file.write("".join(hex_values) + "\n")
+                if self.DebugMode:
+                    file.write("-" * (int(self.NozzleNum/4)) + "\n") # line separator between layers
+
+        # # Visualize voxels with pyvista
+        # def visualize_voxels(voxel_centers):
+        
+        #     # Create a point cloud representation
+        #     point_cloud = pv.PolyData(voxel_centers) #pv.PointCloud(voxel_polydata)
+
+        #     # Plot the point cloud using pyvista plotter
+        #     point_cloud.plot(eye_dome_lighting=True)
+        # visualize_voxels(voxel_array)
+
+        ### Print some data for the user to file
+        orig_stdout = sys.stdout
+        f = open(log_file_path, "w")
+        sys.stdout = f
+        print("-- Printing and slicing statistics --")
+        # Number of voxels
+        tot_voxel_num = np.count_nonzero(voxel_3d_array)
+        print(f"Number of voxels: {tot_voxel_num}")
+        # Total voxel volume
+        tot_voxel_volume = tot_voxel_num * self.VoxelDimX * self.VoxelDimY * self.VoxelDimZ
+        print(f"Print volume: {tot_voxel_volume*1000:.3f} liters")
+        # Number of layers
+        print(f"Number of layers: {unique_z.shape[0]}")
+        # Requested voxel dimensions
+        print("Requested voxel dimensions:")
+        print(f"\t x: {self.VoxelDimX*1000:.3f} mm")
+        print(f"\t y: {self.VoxelDimY*1000:.3f} mm")
+        print(f"\t z: {self.VoxelDimZ*1000:.3f} mm")
+        # Actual voxel dimensions
+        warning_x = ""
+        warning_y = ""
+        warning_z = ""
+        target_deviation = 0.05
+        if np.abs(1 - sliced_voxel_dim_x/self.VoxelDimX) > target_deviation:
+            warning_x = " \t -> Voxel size deviation threshold reached"
+        if np.abs(1 - sliced_voxel_dim_y/self.VoxelDimY) > target_deviation:
+            warning_y = " \t -> Voxel size deviation threshold reached"
+        if np.abs(1 - sliced_voxel_dim_z/self.VoxelDimZ) > target_deviation:
+            warning_z = " \t -> Voxel size deviation threshold reached"
+        print("Actual voxel dimensions:")
+        print(f"\t x: {sliced_voxel_dim_x*1000:.3f} mm" + warning_x)
+        print(f"\t y: {sliced_voxel_dim_y*1000:.3f} mm" + warning_y)
+        print(f"\t z: {sliced_voxel_dim_z*1000:.3f} mm" + warning_z)
+        sys.stdout = orig_stdout
+        f.close()
 
+    def load_standard(self, std: str = None):
+        """Load standard config file
+
+        :param std: defaults to None
+        :type std: str, optional
+        :raises ValueError:
+        """
+        directory = os.path.join(os.path.dirname(os.path.abspath(__file__)), "config")
```

Comment:
this will not work in github actions you need to use the ROOT_PATH see load_standard fct of gcode from points

that's the same function as in Yuxiangs gcode writer correct? Can we move the function to the parent class (gcode)?

Reply:
I moved the load_standard fct to the base class. you can access that now by self.load_standard(standard).